sdlib/d/gc/tcache.d

```diff
@@ -51,9 +64,74 @@ public:
 		}
 
 		auto pd = getPageDescriptor(ptr);
+		// Clear metadata:
+		if (!pd.isSlab())
+			pd.extent.clearLargeMeta();
+
 		pd.arena.free(emap, pd, ptr);
 	}
 
+	// Determine whether given alloc is appendable
+	bool is_appendable(void* ptr) {
+		if (ptr is null)
+			return false;
+
+		auto pd = getPageDescriptor(ptr);
+		// Not supports slab allocs yet
+		if (pd.isSlab())
+			return false;
+
+		return pd.extent.appendable;
+	}
+
+	// Get the current fill of an appendable alloc.
+	// If the alloc is not appendable, returns 0.
+	ulong get_appendable_fill(void* ptr) {
+		if (ptr is null)
+			return 0;
+
+		auto pd = getPageDescriptor(ptr);
+		// Not supports slab allocs yet
+		if (pd.isSlab() || !pd.extent.appendable)
+			return 0;
+
+		return pd.extent.allocSize;
+	}
+
+	// Get the current free space of an appendable alloc.
+	// If the alloc is not appendable, returns 0.
+	ulong get_appendable_free_space(void* ptr) {
+		if (ptr is null)
+			return 0;
+
+		auto pd = getPageDescriptor(ptr);
+		// Not supports slab allocs yet
+		if (pd.isSlab() || !pd.extent.appendable)
+			return 0;
+
+		return cast(ulong) pd.extent.size - pd.extent.allocSize;
+	}
+
+	// Change the current fill of an appendable alloc.
+	// If the alloc is not appendable, or the requested
+	// fill exceeds the available space, returns false.
+	bool set_appendable_fill(void* ptr, ulong fill) {
+		if (ptr is null)
+			return false;
+
+		auto pd = getPageDescriptor(ptr);
+		// Not supports slab allocs yet
+		if (pd.isSlab())
+			return false;
+
+		if (fill <= pd.extent.size) {
+			pd.extent.allocSize = fill;
+			return true;
+		}
+
+		return false;
+	}
```

Comment:
This API has numerous problems, a number of which would have easily been detected if we had tests for it. For instance, what happens when it is called on a pointer that isn't managed by the GC? Or what about interior pointers?

Consider:
```d
uint[] slice = (new uint[123])[0 .. 123];
slice = slice[10 .. $];
slice ~= 42;
```

If there is room, there is no reason for this to reallocate.

Having tests that exercise the API to achieve this kind of result would tell you if the current API is the right one or not. I predict there will be superfluous methods in there, and that the remaining one will have to be amended substantially.